### PR TITLE
feat(0.13.9): metadata 條件標籤篩選（pill）

### PR DESCRIPTION
**Pull request by slive777**
**儀式強度：T3** — 理由：純前端篩選 UX。最壞情況是 localStorage 的 pill 狀態壞掉導致封面牆篩空，清一次就好。不碰影片檔／NFO／DB migration／憑證／批次寫入路徑，安全邊界無涉（未動任何 gate／CSRF／SSRF）。

依 `pre-merge.md` 步驟 0 的分級表，**3.6 來源金絲雀、3.8 跨廠牌 reviewer、契約真理表、push 前 red-team、owner 主持 e2e 皆刻意未跑**，不是漏做。mutation 自驗 T3 不要求，但本支十個 task 每個都做了（理由見下）。

---

## 這支在做什麼

Showcase 的篩選從「一次一個關鍵字」變成「可以疊加的條件」。

以前在燈箱點一個女優，搜尋框內容被整段換掉——想同時看「這個女優 ＋ 這個片商」做不到，點第二個就把第一個洗掉。現在每點一次長一枚可移除的 pill，六種維度（女優／標籤／片商／導演／系列／廠牌）都可以，多枚取交集，自由打字的關鍵字也能與 pill 並存一起參與交集。

**pill 是精準比對，打字仍是模糊比對**——點「巨乳」只出現標籤剛好是「巨乳」的片，不會撈進「巨乳痴女」；但自己打「巨乳」照舊撈到兩者。點進來的是畫面上那一個，多給別的就是錯的。

## 十個 task

| Task | 內容 |
|---|---|
| T1 | pill 狀態層（`addPill`／`removePill`／`clearAllFilters`） |
| T2 | 精準比對層（六維度 + alias 展開） |
| T3 | 持久化（`$persist` 往返 + 舊格式容錯） |
| T4 | 入口改道（十個 metadata 入口）＋ 守衛等價遷移 |
| T5 | pill UI 殼（`x-for` + 複合 key + a11y 清單語意） |
| T6 | pill 視覺、24px 命中區、元件登記 + 修行動版工具列被切 |
| T7 | 清除鈕收斂為單一擁有者 ＋ `showcaseHasSearch` 涵蓋 pill |
| T8 | hero card 收成單一判斷點（七個呼叫點） |
| T9 | Backspace 刪最後一枚 pill（IME 組字中一律不刪） |
| T10 | 狀態列文案「符合 N 個條件的 M 部」 |

`web/static/js/shared/pill-filter.js` 是新模組，為正規化／比對／序列化的單一所有者（零 import 純函式）。

## 幾個值得 reviewer 注意的決策

- **兩層分名**：`searchFromMetadata` 保留原名並降格為只管燈箱關閉生命週期的 adapter，core 是 `addPill`。兩者若同名，`mergeState` 的 `Object.defineProperties` 會讓燈箱那份**靜默蓋掉**核心那份（`main.js` 合併順序 stateVideos 第 2、stateLightbox 第 4），沒有 error。
- **`x-for` 用複合鍵** `dim + '::' + normalizePillValue(value)`：spec 明文允許畫面上出現兩枚字面相同的 pill（系列「Madonna」vs 廠牌「Madonna」），用 value 當 key 會讓 DOM diff 把移除錯配到另一枚。
- **未知 dimension fail-closed**：不誤判為「全部符合」。
- **alias 查表用 `hasOwnProperty` + `Array.isArray`**：裸 `groupMap[key]` 會取到 `Object.prototype` 的繼承成員——一個字面叫 `constructor` 的標籤會讓整個 filter 拋錯，症狀是**空白封面牆**。
- **持久化深拷貝**：既有 9 個 persist 欄位全是 primitive，整包賦值從沒遇過參照問題；陣列賦的是參照，不切斷的話後續每次增刪都在原地改 persisted 巢狀陣列。
- **hero card 的 staleness guard 依 source 分流**：照抄既有的 `search.trim() === capturedTerm` 會讓 pill 路徑的卡片**永遠不出現且無任何錯誤訊息**。修法不是把 pill 值寫進 `this.search`。
- **`role="list"` 放內層** `.filter-pill-list` 而非含 `<input>` 的容器（ARIA 容器語意）。
- **placeholder 同時給靜態與綁定屬性**：只給 Alpine 綁定的話，水合前 100% 的載入都沒有 placeholder。
- **grid track 由裸 `1fr` 改 `minmax(0, 1fr)`**：auto-minimum 取 min-content 當下限，長系列名的 pill 是第一個撐得動它的內容（360px 下把工具列撐到 458px，右緣連圓角被螢幕切掉）。**這個缺陷通過了 `scrollWidth <= innerWidth` 斷言**，因為它是被裁掉不是被捲動。

## 驗證

| 項目 | 結果 |
|---|---|
| pytest（全套，改版號後重跑） | **6955 passed, 1 skipped**（與 v0.13.8 同——純前端，未新增 Python 測試，只遷移兩支既有守衛） |
| `npm test` | **468**（v0.13.8 為 330，本支 +138，分十個檔對應十個 task） |
| `npm run lint` | 六段全綠（eslint／stylelint／css-guard／html／i18n／cjk） |
| `ruff check .` ／ `lint-imports` ／ 函式規模閘 | 全庫乾淨 ／ 1 kept 0 broken ／ 1010 函式 PASS，四條架構閘零新增豁免 |
| 敏感掃描 | 21 檔全掃，零真實洩漏 |
| 打包 | `build.py` 與 `build_macos.py` 的 `COPY_ITEMS` 一致，本支只動 `web/`＋`locales/`（已涵蓋） |
| Capabilities | 零 API surface 變更（diff 無 `web/api/`／`core/`／`main.py`），不需同步 |
| 守衛總量三桶棘輪 | **11,416，本支 delta 0** |

**mutation 自驗做了十個 task 全部**（T3 不要求）。理由是這支的測試幾乎全是「讀原始碼字面」型，不自驗就分不出真守衛與綠殼。過程中靠自驗抓到的：一支新守衛被實作者自己的註解觸發（改成比對函式定義形狀 ＋ 剝註解，而不是放寬守衛）；一條 `anyOf` 規則的紅燈在 OR-of-existence 語意下根本不可能出現（修的是任務卡不是守衛）。

**CDP 真機驗收**：pill 增／刪／疊加、Backspace、清除鈕、hero card 收斂、狀態列文案，都在停用快取的硬重新整理後用真的點擊驗過。這支有數個缺陷是「單元測試全綠但功能是壞的」——除了上面那條 360px 截斷，還有狀態列渲染出 `[showcase.status.pill_prefix]` 字面（查明是開發伺服器的翻譯 `@lru_cache` 沒有 mtime 失效，重啟即消，**不是產品缺陷**，故意沒有為它加 mtime 檢查）。

**dim 主題**由 owner 真機確認通過（pill 細框可辨、填色比 bar 亮一階）。

---

## 已接受的 residual

owner 已完成一輪靜態 code review，未發現 P0/P1 或需修正的產品行為問題。

| # | 項目 | 裁決 |
|---|------|------|
| 1 | 360px 下長 pill 截斷到約 5 字，手機無 hover 看不到 tooltip 全值 | 接受為 P3，列 backlog（窄螢幕 bottom-sheet 或收窄 spotlight 保留槽） |
| 2 | 六個 `pill-*.test.mjs` 以 node:test 文字解析做結構斷言，少數屬「某字面不應出現」型 | 不列為本 branch finding；CD-10 已選擇行為測試而非新增 lint。是否全庫統一機制另案 |
| 3 | `applyFilterAndSort()` 在 pills 為空時仍走一次 `_videos.filter(() => true)` | **不修**。修法（`pills.length === 0 ? _videos : ...`）會讓 `_filteredVideos` 變回 `_videos` 的同一個參考，而下一步 `_filteredVideos.sort()` 是**原地排序**，等於把「排序順手重排主庫陣列」放回來。是行為改變不是簡化；省下的是一次 `return true` 的走訪，緊接著就有一次比較器很重的 O(n log n) 排序無條件在跑 |
| 4 | `.filter-pill-group` 把 `theme.css` 的 spotlight 框逐值複製一份 | **不修**。deeper fix（抽 `.spotlight-frame` 共用類）要動 `theme.css` ＋ `search.html`，會讓本支從未碰過的 search 頁進入視覺回歸風險。兩處差異僅 `height` vs `min-height`（wrapper 要能隨 pill 換行長高）。已在 `theme.css` 補交叉引用註解閉合漂移路徑 |
| 5 | `source === 'metadata'` 已無生產呼叫點 | **不修**。不是遷移遺留垃圾，是 T8 卡要求的等價保存，`pill-hero.test.mjs:308-312` 有守衛明文鎖「`'manual'`/`'metadata'` 逐位元組不變」 |
| 6 | `_reconcileHeroCard()` 不是完整的單一判斷點——`toggleActressMode()` 進女優模式仍直接呼叫 `_clearPreciseMatch()` | **不修，但補註解**。`_shouldShowHeroCard()` 不判斷 `showFavoriteActresses`，該 bypass 是刻意的；原註解寫「七個觸發點全部間接透過本方法」會誘導未來在女優牆可達的程式碼無條件呼叫它（症狀：女優牆上冒出一張不該出現的資料卡）。已補「⚠ 一個刻意的例外」段並寫下正解 |

`/simplify` 跑了（diff 性質＝新模組），四個角度各一支 agent 共提 9 條，**0 採納**——每條問「這條不修，真實使用者會看到、失去、洩漏或必須重做什麼」，答案都是「今天什麼都不會」。其中兩條的失效模式（誤導未來的人）以純註解閉合，零行為變更。

## 其他

- **lint-guard 標註**：T4 遷移的兩支既有守衛帶四條 `assert "<字面>" in js_text`，依步驟 5.7 在範圍內而從未標註。判 `pytest-justified` 不搬 lint——兩條守的是**跨檔關係**而非單檔字面（① adapter 交棒給 `addPill` ② state-base 寫入 → `store.ui.showcaseHasSearch` → 模板讀取），`static_guard_lint` 的 required-string 一次只綁一個檔，拆兩條會失去連結。
- **帳本更正**：`pre-merge.md` 記載的 v0.13.8 棘輪基準 11,412 與實測差 4 行，在 main 上實測為 **11,416**（`wc -l` 與 `grep -c ''` 一致）。非本支造成。
- **i18n**：新 key 只寫 `zh_TW`，其餘三語留空走 fallback（milestone 統一補），`i18n_lint` 的三則 ⚠ 是預期內的。

---

## Review 輪次與追加的 residual（2026-08-10）

三輪 Codex review，**採納 2 條、不採納 3 條**，另修掉 1 條 CI 紅燈（非 Codex 提出）。

### 已修

| 輪 | 條目 | 處置 |
|---|---|---|
| 1 | P1 `test_showcase_mobile_search.py` extractor 錨在裸字串 | **修**（`55e52571`）。但 Codex 的「fails unconditionally」宣稱**不成立**——`state-lightbox.js:467` 那句註解沒有左括號，測試實跑 14 passed。脆弱性屬實且只差一個字元，改為錨定方法定義形狀 |
| — | **CI `lint-frontend` 紅燈**（Codex 未提，owner 指出） | **修**（`a0bce08b`）。`pill-visual.test.mjs` 在 module 頂層讀 gitignored 的 `feature/AI_COLLABORATION/ui-conventions.md`，CI checkout 沒有該檔 → ENOENT → 整支 7 支測試 exit 1。改查已追蹤的 `design-system.html`（依 CLAUDE.md「UI 元件規則」本來就是正典登記處）。驗證方式同時改成 `git ls-files \| tar` 出只有已追蹤檔案的樹再跑 |
| 2 | P3 捲動自動收合守衛漏掉 pill | **修**（`826e249b`）。**本支自引回歸**：pill 之前 `showcaseHasSearch` 與該守衛的判準一致，T7 只放寬前者。改用 `_hasActiveFilter()`，配剝註解的回歸鎖 |

### 已接受的 residual（追加）

| # | 項目 | 裁決 |
|---|------|------|
| 7 | **手機 ≤480px：從燈箱點 metadata（或重整還原 pill）後，工具列是關的而 navbar 鈕已變 ✕，看不到也無法單獨移除該枚 pill，只能全清** | **不修（owner 裁決）**。實機確認平板尺寸正常，僅 ≤480px 直向出現。根因是那顆鈕的職責由「有沒有篩選」決定（`base.html:502-505`）——pill 之前安全，因為「有篩選」只可能發生在工具列開著並打了字之後。根治要翻 owner 2026-06-22 拍板的 81-T2 手機行為，為一個可回復狀態（全清後重點一次）翻既有拍板，代價大於收益 |
| 8 | 重整後單一女優 pill 不會自動顯示女優資料卡，需再互動一次 | **不修——非本 PR 引入**。`main` 上 `search` 本來就在 `$persist`、`restoreState()` 會還原，而 `init()` 從來沒呼叫過 hero 判斷（`main:314` 既有註解自承 restore 不觸發 watcher）。pill 之前用精準命中女優名的搜尋字重整，症狀完全相同。列 backlog；真要修必須同時把 `!showFavoriteActresses` 併進 `_shouldShowHeroCard()` |
| 9 | 重新刮削改掉被篩選的欄位後，不符條件的卡仍留在結果與計數中 | **不修（owner 裁決，edge case）——非 pill 引入**。`refreshVideoData()` 一直是就地 `Object.assign` 不重跑篩選，自由文字搜尋時症狀相同。代價是多看到一張卡，任何篩選動作或重整即恢復 |

### CODE REVIEW

File: tests/unit/test_showcase_mobile_search.py [hunk 1]
```diff
@@ -83,10 +84,61 @@ def test_showcase_has_window_listener(self):
         content = self._read_showcase_html()
         assert "showcase:clear-search" in content
 
-    def test_search_from_metadata_sets_this_search(self):
-        """searchFromMetadata 必須仍設 this.search（防止重構靜默斷開 showcaseHasSearch 路徑）"""
+    @staticmethod
+    def _extract_search_from_metadata_body(content):
+        """brace-matched 擷取 searchFromMetadata 函式體（非整檔 grep，防 FE-GUARD-06 假陽性——
+        `this.search = ` 這個字面在檔案其他函式仍合法存在）。"""
+        sig_idx = content.find("searchFromMetadata(")
```
[the remark this answers is not in the document]
Comment (slive777) [reply tests/unit/test_showcase_mobile_search.py]: 已修（55e52571），但先更正一件事：**「fails unconditionally」不成立。**

`state-lightbox.js:467` 那行是 `// F2: cancel pending delayed clear from previous close / searchFromMetadata`——**沒有左括號**，而 extractor 找的是 `"searchFromMetadata("`。該測試實跑 `1 passed`，整支檔案 `14 passed`。

不過方向對，而且只差一個字元。把兩個 extractor 各餵「原始檔案」與「注入一行帶括號註解的檔案」，實測：

| 情境 | 舊裸 `find` | 新 regex |
|---|---|---|
| 今天的檔案 | 正確 body（1218 字元，含 `addPill(`） | 同（1218） |
| 注入 `// 見 searchFromMetadata(value, dim) 的關閉流程` | `closeLightbox()` 的 body（4730 字元，無 `addPill(`）→ 守衛假紅 | 仍是正確的 1218 |

已改成 `re.search(r"searchFromMetadata\s*\([^)]*\)\s*\{")`，與同檔的 `_extract_has_active_filter_body` 一致（它本來就錨定定義形狀）。docstring 記下 467 行那句註解已經在檔案裡，讓下一個人知道這不是假想風險。

pytest 全套 6955 passed, 1 skipped，無變化。

順帶：掃了全庫同型錨定，`tests/unit/test_frontend_lint.py:3463` 錨在裸 `"cf_unavailable"`，實際命中 `state-rescrape.js:228` 的註解而非 238 行的 `if (data && data.cf_unavailable)`——今天巧合通過，是潛在假綠。非本 PR 改動範圍，另案處理。

**Comment by slive777** [comment]
@codex review

File: web/static/js/pages/showcase/state-base.js [hunk 2]
```diff
@@ -326,12 +334,9 @@ export function stateBase() {
...
-        clearSearch() {
-            this.search = ''
-            this.actressSearch = ''
-            this.onSearchChange()
-            this.onActressSearchChange()
-            Alpine.store('ui').toolbarOpen = false
+        // 115-T7 / CD-12：單一「是否有啟用中篩選」判準（文字／女優文字／pill 任一即真）
+        _hasActiveFilter() {
+            return this.search !== '' || this.actressSearch !== '' || this.pills.length > 0;
         },
```
[the remark this answers is not in the document]
Comment (slive777) [reply web/static/js/pages/showcase/state-base.js:348]: 已修（826e249b）。查證屬實，而且**這是本支自己引入的回歸**，依 CLAUDE.md「本次改動自己引入的回歸也必須被處理」修，不當 P3 放掉。

使用者流程走得通，我逐段核過：手機（≤480px）只用 pill 篩選、沒打任何字 → 往下捲超過 50px → 裝著 pill 的整條工具列被收掉 → navbar 那顆鈕因 `showcaseHasSearch`（T7 起含 pills）為真而顯示 ✕，而 `base.html:502-505` 是 `showcaseHasSearch ? $dispatch(clear-search) : (toolbarOpen = !toolbarOpen)`——按下去是**全清，不是重新展開**。所以使用者再也無法只移除其中一枚 pill。

**為什麼是自引**：pill 之前 `showcaseHasSearch` 只在有搜尋字時為真，而本守衛的 `search !== ` 剛好蓋住同一個情況，兩個判準原本一致。T7 把 `showcaseHasSearch` 放寬成 `_hasActiveFilter()`（CD-12），本守衛沒跟上。同段既有註解自己寫著「搜尋空白時」才收合——有 pill 就不是空白。

修法是 `if (this.search !== "" || this.actressSearch !== "") return` → `if (this._hasActiveFilter()) return`（嚴格超集，只多 `pills.length > 0`）。

配了回歸鎖：brace-match 抽 `_scrollHandler` body、**剝註解後**才比對（否則新加的說明註解自己會讓守衛通過），正向要求 `_hasActiveFilter()`、反向禁止舊的兩欄位字面；mutation 兩個方向各只讓一支轉紅。

另外同一輪修掉了 CI 紅燈（a0bce08b，與本條無關）：`pill-visual.test.mjs` 在 module 頂層讀 gitignored 的 `feature/AI_COLLABORATION/ui-conventions.md`，CI checkout 沒有該檔 → ENOENT → 整支檔案 7 支測試 exit 1。已改查已追蹤的 `design-system.html`。驗證方式也改成真的 CI 條件（`git ls-files | tar` 出只有已追蹤檔案的樹再跑）：npm test 469 pass / 0 fail、npm run lint exit 0、pytest 6955 passed。

**Comment by slive777** [comment]
@codex review

File: web/static/js/pages/showcase/state-lightbox.js [hunk 3]
```diff
@@ -675,13 +680,7 @@ export function stateLightbox() {
...
             }, 250);
 
-            this.search = term;
-            this._animateFilter();
-            if (type === 'actress') {
-                this._checkPreciseActressMatch(term, 'metadata');
-            } else {
-                this._clearPreciseMatch();
-            }
+            this.addPill(dim, value);
```
[the remark this answers is not in the document]
Comment (slive777) [reply web/static/js/pages/showcase/state-lightbox.js:683]: **不修（owner 裁決）。**

查證屬實：手機（≤480px）從燈箱點 metadata 後 pill 有加進去，但工具列仍是關的，而 navbar 那顆鈕因 `showcaseHasSearch` 為真而變成 ✕（`base.html:502-505` 是三元式：有篩選就 dispatch `clear-search`，不再 toggle 工具列），所以看不到也無法單獨移除該枚 pill。另有一條你沒提到但同源的路徑：手機上重新整理，還原的 pill 讓 `showcaseHasSearch` 為真而 `toolbarOpen` 預設 false，一載入就是同樣狀態。

根因不在這個 call site，而在那顆鈕的職責被「有沒有篩選」決定——pill 之前這是安全的，因為「有篩選」只可能發生在「工具列開著並打了字」之後。

**owner 裁決不修**：實機確認平板尺寸正常，此死路只在 ≤480px 手機直向出現，而根治法要改 `base.html` 那顆鈕的語意，那是 owner 於 2026-06-22 拍板的既有手機行為（81-T2）。為一個窄尺寸的可回復狀態（全清後重點一次即可）去翻既有拍板，代價大於收益。列為已接受的 residual。

File: web/static/js/pages/showcase/state-base.js [hunk 4]
```diff
@@ -362,6 +375,7 @@ export function stateBase() {
             this.perPage = defaultPerPage;
             this.page = parseInt(urlNum('page') ?? state.page ?? 1) || 1;
             this.search = urlParams.get('search') || state.search || '';
+            this.pills = deserializePills(state.pills);
```
[the remark this answers is not in the document]
Comment (slive777) [reply web/static/js/pages/showcase/state-base.js:378]: **不修——這是 pill 之前就存在的行為，非本 PR 引入。**

在 `main` 上核過：`search` 本來就在 `$persist` 清單內，`restoreState()` 會還原它（`main:364`），而 `init()` 的順序是 `restoreState()` → `applyFilterAndSort(true)`，**從來沒有呼叫過 hero card 判斷**；`main:314` 的既有註解自己寫著「restoreState() 在 $watch 前執行，初始 search/actressSearch 不觸發 watcher」。

也就是說 pill 之前：搜尋框留著剛好精準命中某位女優的字 → 重新整理 → 片子篩對了，資料卡一樣不會自己出現，要等下一次互動。症狀與你描述的完全相同。pill 只是換一種方式落進同一個舊坑，沒有讓它變糟。

依本專案「非本 branch 改動不得混入本次修正」，列 backlog 另案處理。順帶記錄：真要修，必須同時把 `!showFavoriteActresses` 併進 `_shouldShowHeroCard()`，否則在女優牆模式下還原一枚女優 pill 會冒出一張不該出現的卡（見 `state-videos.js` `_reconcileHeroCard()` 上方的 ⚠ 註解）。

File: web/static/js/pages/showcase/state-videos.js [hunk 5]
```diff
@@ -299,12 +400,16 @@ export function stateVideos() {
 
         // --- 資料處理 ---
         applyFilterAndSort(skipPagination) {
-            // --- 搜尋篩選 (M4a) ---
+            // --- Stage 1: pill 精準比對（TASK-115-T2, CD-6）---
+            var pillPredicate = buildPillPredicate(this.pills, _nameToGroup, _tagToGroup);
+            var pillFiltered = _videos.filter(pillPredicate);
```
[the remark this answers is not in the document]
Comment (slive777) [reply web/static/js/pages/showcase/state-videos.js:405]: **不修（owner 裁決）——且這不是 pill 引入的。**

`refreshVideoData()`（`state-lightbox.js:1694`）一直都是 `Object.assign(video, data.video)` 就地改資料、不重跑篩選。pill 之前用自由文字搜尋時症狀相同：用標題關鍵字篩選 → 對其中一片重新刮削 → 標題被換掉 → 那張卡照樣留在結果裡、計數照算，直到重新整理或動一下篩選。pill 只是多一種觸發方式，並未加重。

使用者代價：多看到一張已不符條件的卡，任何一次篩選動作或重整即恢復；無資料遺失、無檔案改動。owner 判定為 edge case 不處理。修它要在重刮流程裡插入重跑篩選，屬本 PR 範圍外的路徑。列為已接受的 residual。
